### ISSUE TITLE
Support YCbCr IPixelSources

### ISSUE THREAD

**Issue by iamcarbon** [saucecontrol/PhotoSauce#31]
We're decoding some images directly into YCbCr planes (via ffmpeg and libheif).

It would be handy to avoid the RGB conversions and pass these buffers in directly to MagicScaler.

e.g.

```csharp
public class YCbCrPixelSource(IntPtr yPlanePointer, IntPtr cbPlanePointer, IntPtr crPlanePointer, ChromaSubsampling chromaSubsampling) { } 

```

**Comment by saucecontrol** [comment]
It just happens I was recently working on simplifying my pixel buffering logic in the pipeline to allow alternate ways of consuming planar data (e.g. for turbojpeg integration), and this fits in nicely with that plan.

Once that buffering rewrite is done, we'll have to add support for separate Cb and Cr planes in addition to the interleaved CbCr format the pipeline currently uses.  WIC supports decoding and encoding in both formats, but limitations with `Vector<T>` meant that it was advantageous to consume the chroma planes interleaved, so that's all I built out.  The new HWIntrinsics will allow building optimized processing for either format.

I'm assuming you're talking about extracting video thumbnails for this use case?  One challenge there is that the chroma alignment (siting) used when subsampling is different depending on video format.  JPEG and MPEG-1 take the chroma sample from the center of the luma samples (interstitial siting), whereas MPEG-2 defined the chroma samples as left-aligned in the horizontal direction (co-sited) and centered in the vertical direction.  H.264 and H.265 use the MPEG-2 siting by default but an encoder can override that with its own setting if desired.  In most cases, interpreting the alignment incorrectly doesn't make a visible difference, but there are certainly cases where it does.  The logic in place now assumes JPEG alignment, so that may need to be expanded, and we may need a chroma alignment property on the pixel source.

**Comment by saucecontrol** [comment]
I've got the interfaces defined and plumbed in to make this work, and it's available now in the [CI builds](https://dev.azure.com/saucecontrol/PhotoSauce/_packaging?_a=feed&feed=photosauce_ci)

I put together a sample using https://github.com/carbon/Heif that shows all the pieces together, which you can find in this gist: https://gist.github.com/saucecontrol/ae60feac2e7c2343aad33556c3e64d47

The main points are the new `IImageContainer` and `IYccImageFrame` interfaces which expose the pixels and mandatory metadata.  I also added a new `BitmapPixelSource` base class that makes it easier to wrap an already decoded in-memory bitmap, and used that in the example to wrap the 3 image planes and expose them as `IPixelSource`s.

Testing that example code with the sample `camel.heic` image in the codec's test folder, I'm getting correct-ish output, but there's still work to be done.  That image makes for an interesting test case because its YCbCr format isn't JPEG compatible, which makes the output levels incorrect when passed though directly to the JPEG encoder or converted to RGB through WIC's PlanarFormatConverter.  You can see when running the same file through MagicScaler with the WIC HEIF codec that the output image is more vibrant in comparison.

Even more interestingly, `camel.heic` has no ICC color profile and no HEIF 'nclx` tag that would describe how to interpret the YCC data, and yet WIC decodes and converts it to BGR correctly.  I haven't been able to tell whether it's getting the info from the h.265 VUI descriptor (which isn't exposed by libheif) or whether it simply defaults to Rec. 709 with video levels (16-235) in the absence of any better information.

In any case, to make this work properly, I'll need to write my own YCC to BGR interleaver/converter and a converter to map between different YCC formats, and we'll need to have the metadata present on the `IYccImageFrame` to control that conversion.  I believe I've set up all the properties I'll need on the interface, but I won't know for sure until they're plumbed in on the codec side and I get my converters done.

Anyway, let me know when you get a chance to look it over and of course let me know if you see any issues or anything I missed.

**Comment by saucecontrol** [comment]
Ok, I've finished plumbing all the `IYccImageFrame` properties through, and it's all working with any `.heic` files I've tried, provided I dummy up the correct metadata.  I tweaked the API a bit and updated the [sample gist](https://gist.github.com/saucecontrol/ae60feac2e7c2343aad33556c3e64d47) accordingly.

The properties set in the sample work for the 2 test images in the libheif wrapper project.  The results match what I get from the WIC codec, and it's 2.5-3x as fast.

However....

I've got some sample `.heic` files from IOS, and they're quite different.  They're encoded with full range values and the Rec. 601 matrix.  Plus they have an embedded Display P3 ICC profile.  If I set all those property values correctly on the sample `IYccImageFrame`, the outputs also match the WIC codec.  But those decode around 25% slower with libheif than with WIC.  It could be libheif is making a round-trip through RGB and back to planar on decode for some reason.

Anyway, the wrapper obviously needs to be updated to fill in the metadata values properly, but I'm not quite sure how some of them should work.  As I suspected, the sample images are encoded with video levels and a Rec 709 matrix, but I'm not certain how the WIC codec is determining that.  If I had to guess based on the few `.heic` files I've seen, I'd define the logic like this:

1. If the file has a `colr` FourCC tag, assume JPEG compatibility (full range values and Rec 601 matrix), plus extract the embedded ICC profile.
2. If no `colr`, check for `nclx`.  That tag has explicit values for range and matrix https://github.com/strukturag/libheif/blob/master/libheif/heif.cc#L828-L841.
3. If neither of those, I'd look for VUI data in the embedded video header.  libheif doesn't appear to expose that anywhere, but I can see it's been set by the x265 encoder used to produce the sample images, so it's in there somewhere.  https://x265.readthedocs.io/en/default/cli.html#vui-video-usability-information-options
4. Absent, that, I guess assume video range and Rec 709 matrix, which are the defaults for AVC and HEVC.

That's all I've worked out for now.  I'll come back around to the metadata and the performance oddities when I have more time to play around with native codecs in gereral.
